File: backend/src/router/migrate.js
const fs = require("fs");
const path = require("path");
const { PassThrough } = require("stream");

const DEFAULT_SOURCE_DIR = process.env.SOURCE_DIR || "/vol1/1000";
const SOURCE_MAP = {
  personal: { dir: "/vol1/1000", label: "personal" },
  team: { dir: "/vol1/@team", label: "team" },
};

function normalizeBaseUrl(baseUrl) {
  return String(baseUrl || "").trim().replace(/\/+$/, "");
}

function normalizeStorage(storage) {
  return String(storage || "").trim().replace(/^\/+|\/+$/g, "");
}

async function readJson(res) {
  try {
    return await res.json();
  } catch {
    return null;
  }
}

function assertResponse(res, data, actionLabel) {
  if (!res.ok) {
    throw new Error(`${actionLabel}失败(${res.status})`);
  }

  if (data && Object.prototype.hasOwnProperty.call(data, "success")) {
    if (data.success !== true && data.success !== 200) {
      throw new Error(data.message || data.msg || `${actionLabel}失败`);
    }
  }

  if (data && Object.prototype.hasOwnProperty.call(data, "code")) {
    if (data.code !== 200) {
      throw new Error(data.message || data.msg || `${actionLabel}失败`);
    }
  }
}

function resolveSource(sourceType) {
  if (sourceType && SOURCE_MAP[sourceType]) {
    return SOURCE_MAP[sourceType];
  }

  if (sourceType) {
    return null;
  }

  if (DEFAULT_SOURCE_DIR === SOURCE_MAP.team.dir) {
    return SOURCE_MAP.team;
  }

  if (DEFAULT_SOURCE_DIR === SOURCE_MAP.personal.dir) {
    return SOURCE_MAP.personal;
  }

  return { dir: DEFAULT_SOURCE_DIR, label: "custom" };
}

function toPosixPath(input) {
  return String(input || "").split(path.sep).join("/");
}

function isDirExistsMessage(message) {
  const msg = String(message || "").toLowerCase();
  return msg.includes("exist") || msg.includes("exists") || msg.includes("已存在") || msg.includes("already");
}

function sendEvent(stream, type, data) {
  const payload = data ? JSON.stringify(data) : "";
  stream.write(`event: ${type}\n`);
  stream.write(`data: ${payload}\n\n`);
}

async function login(baseUrl, username, password) {
  const loginRes = await fetch(`${baseUrl}/v1/users/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username, password }),
  });
  const loginData = await readJson(loginRes);
  assertResponse(loginRes, loginData, "登录");

  const token =
    loginData?.data?.token?.access_token ||
    loginData?.data?.token ||
    loginData?.token ||
    "";

  if (!token) {
    throw new Error("登录成功但未获取到 token");
  }

  return token;
}

async function scanSourceDir(rootDir) {
  const files = [];
  const dirs = [];
  const stack = [""];

  while (stack.length) {
    const relDir = stack.pop();
    const absDir = path.join(rootDir, relDir);
    const entries = await fs.promises.readdir(absDir, { withFileTypes: true });

    for (const entry of entries) {
      const relPath = path.join(relDir, entry.name);
      if (entry.isDirectory()) {
        dirs.push(relPath);
        stack.push(relPath);
      } else if (entry.isFile()) {
        files.push(relPath);
      }
    }
  }

  return { files, dirs };
}

async function ensureRemoteDir({ baseUrl, token, dirPath, supportsMkdir }) {
  if (!supportsMkdir.value || !dirPath) {
    return;
  }

  const res = await fetch(`${baseUrl}/v2_1/files/folder`, {
    method: "POST",
    headers: {
      Authorization: token,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ path: dirPath }),
  });
  const data = await readJson(res);

  if (res.status === 404) {
    supportsMkdir.value = false;
    return;
  }

  if (!res.ok) {
    const message = data?.message || data?.msg || "";
    if (isDirExistsMessage(message)) {
      return;
    }
    throw new Error(message || `创建目录失败(${dirPath})`);
  }

  if (data && Object.prototype.hasOwnProperty.call(data, "success")) {
    if (data.success !== true && data.success !== 200) {
      if (!isDirExistsMessage(data.message || data.msg)) {
        throw new Error(data.message || data.msg || `创建目录失败(${dirPath})`);
      }
    }
  }

  if (data && Object.prototype.hasOwnProperty.call(data, "code")) {
    if (data.code !== 200) {
      if (!isDirExistsMessage(data.message || data.msg)) {
        throw new Error(data.message || data.msg || `创建目录失败(${dirPath})`);
      }
    }
  }
}

async function uploadFile({ baseUrl, token, remoteDir, filename, fullPath }) {
  const [fileBuffer, stat] = await Promise.all([
    fs.promises.readFile(fullPath),
    fs.promises.stat(fullPath),
  ]);
  const formData = new FormData();
  formData.append("path", remoteDir);
  formData.append("modTime", Math.floor(stat.mtimeMs / 1000).toString());
  formData.append(
    "file",
    new Blob([fileBuffer], { type: "application/octet-stream" }),
    filename
  );

  const uploadRes = await fetch(`${baseUrl}/v2_1/files/file/uploadV2`, {
    method: "POST",
    headers: { Authorization: token },
    body: formData,
  });
  const uploadData = await readJson(uploadRes);
  assertResponse(uploadRes, uploadData, "上传");
}

async function runMigration(stream, options) {
  const { baseUrl, username, password, storagePath, sourceInfo } = options;
  const dstPath = storagePath;
  let currentStep = "login";

  try {
    sendEvent(stream, "progress", {
      step: "login",
      status: "start",
      message: "正在登录 ZimaOS...",
    });
    const token = await login(baseUrl, username, password);
    sendEvent(stream, "progress", {
      step: "login",
      status: "success",
      message: "登录成功",
    });

    currentStep = "scan";
    sendEvent(stream, "progress", {
      step: "scan",
      status: "start",
      message: "正在扫描目录...",
    });
    const { files, dirs } = await scanSourceDir(sourceInfo.dir);
    const totalFiles = files.length;
    sendEvent(stream, "progress", {
      step: "scan",
      status: "success",
      message: `扫描完成：${totalFiles} 个文件`,
      totalFiles,
    });

    currentStep = "upload";
    const supportsMkdir = { value: true };
    const createdDirs = new Set();
    const sortedDirs = dirs.map(toPosixPath).sort((a, b) => a.length - b.length);

    for (const relDir of sortedDirs) {
      if (!relDir || createdDirs.has(relDir)) {
        continue;
      }
      const remoteDir = path.posix.join(dstPath, relDir);
      await ensureRemoteDir({ baseUrl, token, dirPath: remoteDir, supportsMkdir });
      createdDirs.add(relDir);
    }

    sendEvent(stream, "progress", {
      step: "upload",
      status: "start",
      message: totalFiles ? "开始上传文件..." : "无需上传文件",
      totalFiles,
    });

    for (let index = 0; index < files.length; index += 1) {
      const relPath = files[index];
      const relPosix = toPosixPath(relPath);
      const fullPath = path.join(sourceInfo.dir, relPath);
      const dirName = path.posix.dirname(relPosix);
      const remoteDir = dirName === "." ? dstPath : path.posix.join(dstPath, dirName);
      const filename = path.posix.basename(relPosix);

      sendEvent(stream, "progress", {
        step: "upload",
        status: "start",
        message: `正在上传 ${index + 1}/${totalFiles}`,
        currentFile: relPosix,
        transferredFiles: index,
        totalFiles,
      });

      await uploadFile({ baseUrl, token, remoteDir, filename, fullPath });
      sendEvent(stream, "progress", {
        step: "upload",
        status: "start",
        message: `正在上传 ${index + 1}/${totalFiles}`,
        currentFile: relPosix,
        transferredFiles: index + 1,
        totalFiles,
      });
    }

    sendEvent(stream, "progress", {
      step: "upload",
      status: "success",
      message: "文件上传完成",
      transferredFiles: totalFiles,
      totalFiles,
    });

    sendEvent(stream, "done", {
      message: "迁移完成",
      result: {
        dstPath,
        sourceDir: sourceInfo.dir,
        sourceType: sourceInfo.label,
        totalFiles,
      },
    });
  } catch (error) {
    error.step = currentStep;
    throw error;
  }
}

module.exports = async function migrate({ body }) {
  if (typeof fetch !== "function") {
    return { code: 500, msg: "当前环境不支持原生 fetch", data: null };
  }

  if (typeof FormData === "undefined" || typeof Blob === "undefined") {
    return { code: 500, msg: "当前环境缺少 FormData/Blob", data: null };
  }

  const baseUrl = normalizeBaseUrl(body.baseUrl);
  const username = String(body.username || "").trim();
  const password = String(body.password || "");
  const storage = normalizeStorage(body.storage);
  const sourceType = String(body.source || body.space || "").trim();

  if (!baseUrl || !username || !password) {
    return { code: 400, msg: "缺少 baseUrl/username/password", data: body };
  }

  const sourceInfo = resolveSource(sourceType);
  if (!sourceInfo) {
    return { code: 400, msg: "未知的迁移空间", data: body };
  }

  if (!fs.existsSync(sourceInfo.dir)) {
    return { code: 400, msg: "源目录不存在", data: null };
  }

  const sourceStat = fs.statSync(sourceInfo.dir);
  if (!sourceStat.isDirectory()) {
    return { code: 400, msg: "源路径不是目录", data: null };
  }

  const storagePath = storage ? `/media/${storage}` : "/media";
  const stream = new PassThrough();

  setImmediate(() => {
    runMigration(stream, {
      baseUrl,
      username,
      password,
      storagePath,
      sourceInfo,
    })
      .catch((error) => {
        sendEvent(stream, "error", {
          step: error?.step,
          message: error?.message || "迁移失败",
        });
      })
      .finally(() => {
        stream.end();
      });
  });

  return {
    code: 200,
    type: "text/event-stream; charset=utf-8",
    headers: {
      "Cache-Control": "no-cache",
      Connection: "keep-alive",
      "X-Accel-Buffering": "no",
    },
    stream,
  };
};

File: backend-go/internal/handler/migrate.go
package handler

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"ftoz/internal/model"
	"ftoz/internal/service"

	"github.com/gin-gonic/gin"
)

const DefaultSourceDir = "/vol1/1000"

var SourceMap = map[string]model.SourceInfo{
	"personal": {Dir: "/vol1/1000", Label: "personal"},
	"team":     {Dir: "/vol1/@team", Label: "team"},
}

// MigrateHandler 迁移处理器
type MigrateHandler struct {
	zimaClient *service.ZimaOSClient
	scanner    *service.Scanner
}

// NewMigrateHandler 创建迁移处理器
func NewMigrateHandler() *MigrateHandler {
	return &MigrateHandler{
		zimaClient: service.NewZimaOSClient(),
		scanner:    service.NewScanner(),
	}
}

// Handle Gin 处理函数
func (h *MigrateHandler) Handle(c *gin.Context) {
	var req model.MigrateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code": 400,
			"msg":  "请求参数解析失败",
			"data": nil,
		})
		return
	}

	h.handleMigrate(c.Writer, &req)
}

// HandleHTTP 标准 HTTP 处理函数 (用于 CGI)
func (h *MigrateHandler) HandleHTTP(w http.ResponseWriter, r *http.Request) {
	var req model.MigrateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		w.Write([]byte(`{"code":400,"msg":"请求参数解析失败","data":null}`))
		return
	}

	h.handleMigrate(w, &req)
}

func (h *MigrateHandler) handleMigrate(w http.ResponseWriter, req *model.MigrateRequest) {
	// 参数验证
	if err := h.validateRequest(req); err != nil {
		h.writeJSON(w, 400, err.Error(), req)
		return
	}

	// 解析源目录
	sourceInfo := h.resolveSource(req.Source, req.Space)
	if sourceInfo == nil {
		h.writeJSON(w, 400, "未知的迁移空间", req)
		return
	}

	// 验证源目录存在
	stat, err := os.Stat(sourceInfo.Dir)
	if os.IsNotExist(err) {
		h.writeJSON(w, 400, "源目录不存在", nil)
		return
	}
	if !stat.IsDir() {
		h.writeJSON(w, 400, "源路径不是目录", nil)
		return
	}

	// 设置 SSE 响应头
	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.Header().Set("X-Accel-Buffering", "no")

	// 获取 Flusher
	flusher, ok := w.(http.Flusher)
	if !ok {
		h.writeJSON(w, 500, "不支持流式响应", nil)
		return
	}
	flusher.Flush()

	// 执行迁移
	h.runMigration(w, flusher, req, sourceInfo)
}

func (h *MigrateHandler) runMigration(w http.ResponseWriter, flusher http.Flusher, req *model.MigrateRequest, sourceInfo *model.SourceInfo) {
	storagePath := "/media"
	if req.Storage != "" {
		storagePath = "/media/" + strings.Trim(req.Storage, "/")
	}

	currentStep := "login"

	// 1. 登录
	h.sendEvent(w, flusher, "progress", model.ProgressEvent{
		Step:    "login",
		Status:  "start",
		Message: "正在登录 ZimaOS...",
	})

	token, err := h.zimaClient.Login(req.BaseURL, req.Username, req.Password)
	if err != nil {
		h.sendEvent(w, flusher, "error", model.ErrorEvent{
			Step:    currentStep,
			Message: err.Error(),
		})
		return
	}

	h.sendEvent(w, flusher, "progress", model.ProgressEvent{
		Step:    "login",
		Status:  "success",
		Message: "登录成功",
	})

	// 2. 扫描目录
	currentStep = "scan"
	h.sendEvent(w, flusher, "progress", model.ProgressEvent{
		Step:    "scan",
		Status:  "start",
		Message: "正在扫描目录...",
	})

	files, dirs, err := h.scanner.Scan(sourceInfo.Dir)
	if err != nil {
		h.sendEvent(w, flusher, "error", model.ErrorEvent{
			Step:    currentStep,
			Message: err.Error(),
		})
		return
	}

	totalFiles := len(files)
	h.sendEvent(w, flusher, "progress", model.ProgressEvent{
		Step:       "scan",
		Status:     "success",
		Message:    fmt.Sprintf("扫描完成：%d 个文件", totalFiles),
		TotalFiles: totalFiles,
	})

	// 3. 创建远程目录
	currentStep = "upload"
	supportsMkdir := true
	createdDirs := make(map[string]bool)

	// 按路径长度排序，确保父目录先创建
	sortedDirs := h.sortDirsByDepth(dirs)

	for _, relDir := range sortedDirs {
		if relDir == "" || createdDirs[relDir] {
			continue
		}
		relPosix := toPosixPath(relDir)
		remoteDir := storagePath + "/" + relPosix

		if supportsMkdir {
			err := h.zimaClient.CreateDir(req.BaseURL, token, remoteDir)
			if err != nil {
				if strings.Contains(err.Error(), "404") {
					supportsMkdir = false
				} else {
					h.sendEvent(w, flusher, "error", model.ErrorEvent{
						Step:    currentStep,
						Message: err.Error(),
					})
					return
				}
			}
		}
		createdDirs[relDir] = true
	}

	// 4. 上传文件
	startMsg := "开始上传文件..."
	if totalFiles == 0 {
		startMsg = "无需上传文件"
	}
	h.sendEvent(w, flusher, "progress", model.ProgressEvent{
		Step:       "upload",
		Status:     "start",
		Message:    startMsg,
		TotalFiles: totalFiles,
	})

	for i, relPath := range files {
		relPosix := toPosixPath(relPath)
		fullPath := filepath.Join(sourceInfo.Dir, relPath)
		dirName := filepath.Dir(relPosix)

		remoteDir := storagePath
		if dirName != "." {
			remoteDir = storagePath + "/" + dirName
		}
		filename := filepath.Base(relPosix)

		h.sendEvent(w, flusher, "progress", model.ProgressEvent{
			Step:             "upload",
			Status:           "start",
			Message:          fmt.Sprintf("正在上传 %d/%d", i+1, totalFiles),
			CurrentFile:      relPosix,
			TransferredFiles: i,
			TotalFiles:       totalFiles,
		})

		err := h.zimaClient.UploadFile(req.BaseURL, token, remoteDir, filename, fullPath)
		if err != nil {
			h.sendEvent(w, flusher, "error", model.ErrorEvent{
				Step:    currentStep,
				Message: err.Error(),
			})
			return
		}

		h.sendEvent(w, flusher, "progress", model.ProgressEvent{
			Step:             "upload",
			Status:           "start",
			Message:          fmt.Sprintf("正在上传 %d/%d", i+1, totalFiles),
			CurrentFile:      relPosix,
			TransferredFiles: i + 1,
			TotalFiles:       totalFiles,
		})
	}

	h.sendEvent(w, flusher, "progress", model.ProgressEvent{
		Step:             "upload",
		Status:           "success",
		Message:          "文件上传完成",
		TransferredFiles: totalFiles,
		TotalFiles:       totalFiles,
	})

	// 5. 完成
	h.sendEvent(w, flusher, "done", model.DoneEvent{
		Message: "迁移完成",
		Result: model.MigrateResult{
			DstPath:    storagePath,
			SourceDir:  sourceInfo.Dir,
			SourceType: sourceInfo.Label,
			TotalFiles: totalFiles,
		},
	})
}

func (h *MigrateHandler) validateRequest(req *model.MigrateRequest) error {
	req.BaseURL = strings.TrimRight(strings.TrimSpace(req.BaseURL), "/")
	req.Username = strings.TrimSpace(req.Username)
	req.Storage = strings.Trim(strings.TrimSpace(req.Storage), "/")

	if req.BaseURL == "" || req.Username == "" || req.Password == "" {
		return fmt.Errorf("缺少 baseUrl/username/password")
	}
	return nil
}

func (h *MigrateHandler) resolveSource(sourceType, space string) *model.SourceInfo {
	// 兼容 source 和 space 参数
	st := strings.TrimSpace(sourceType)
	if st == "" {
		st = strings.TrimSpace(space)
	}

	if info, ok := SourceMap[st]; ok {
		return &info
	}

	if st != "" {
		return nil
	}

	// 默认源目录
	envDir := os.Getenv("SOURCE_DIR")
	if envDir == "" {
		envDir = DefaultSourceDir
	}

	for _, info := range SourceMap {
		if info.Dir == envDir {
			return &info
		}
	}

	return &model.SourceInfo{Dir: envDir, Label: "custom"}
}

func (h *MigrateHandler) sortDirsByDepth(dirs []string) []string {
	sorted := make([]string, len(dirs))
	copy(sorted, dirs)
	sort.Slice(sorted, func(i, j int) bool {
		return len(sorted[i]) < len(sorted[j])
	})
	return sorted
}

func (h *MigrateHandler) sendEvent(w http.ResponseWriter, flusher http.Flusher, eventType string, data interface{}) {
	payload, _ := json.Marshal(data)
	fmt.Fprintf(w, "event: %s\n", eventType)
	fmt.Fprintf(w, "data: %s\n\n", payload)
	flusher.Flush()
}

func (h *MigrateHandler) writeJSON(w http.ResponseWriter, code int, msg string, data interface{}) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	json.NewEncoder(w).Encode(model.Response{
		Code: code,
		Msg:  msg,
		Data: data,
	})
}

func toPosixPath(p string) string {
	return strings.ReplaceAll(p, "\\", "/")
}
